**Linux** [Build 9655](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=94638&buildTypeId=WFPR) outcome was **FAILURE** using a merge of 6107d6643780aa27c63e93a29eff2df50a74cc5d
Summary: Tests failed: 1 (1 new), passed: 3318, ignored: 194 Build time: 00:59:00
### Failed tests
```
org.jboss.as.test.integration.web.response.DefaultResponseCodeAtRootTestCase.testDefaultResponseCode: java.lang.AssertionError: GET http://[::1]:8080/ HTTP/1.1 expected:<506> but was:<200>
	at org.jboss.as.test.integration.web.response.DefaultResponseCodeAtRootTestCase.testDefaultResponseCode(DefaultResponseCodeAtRootTestCase.java:125)
------- Stdout: -------
 [0m15:41:17,408 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 5) WFLYUT0022: Unregistered web context: /jaxrsnoap
 [0m [0m15:41:17,411 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-4) WFLYSRV0028: Stopped deployment jaxrsnoap.war (runtime-name: jaxrsnoap.war) in 3ms
 [0m [0m15:41:17,414 INFO  [org.jboss.as.repository] (management-handler-thread - 4) WFLYDR0002: Content removed from location /opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/basic/target/jbossas/standalone/data/content/8e/5c1cdaa570ca16c39bbdbe9f8b9d4c3ac2f440/content
 [0m [0m15:41:17,414 INFO  [org.jboss.as.server] (management-handler-thread - 4) WFLYSRV0009: Undeployed "jaxrsnoap.war" (runtime-name: "jaxrsnoap.war")
 [0m [0m15:41:17,549 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0010: Unbound data source [java:jboss/datasources/ds_TestDS]
 [0m [0m15:41:17,549 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0010: Unbound data source [java:jboss/datasources/ExampleDS]
 [0m [0m15:41:17,550 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0010: Unbound data source [java:jboss/datasources/MaskedDS]
 [0m [0m15:41:17,550 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0010: Unbound data source [java:jboss/datasources/DBLMTest]
 [0m [0m15:41:17,550 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0010: Unbound data source [java:jboss/datasources/ds_TestDSWrong]
 [0m [0m15:41:17,551 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-2) WFLYUT0019: Host default-host stopping
 [0m [0m15:41:17,560 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-4) WFLYUT0008: Undertow HTTP listener default suspending
 [0m [0m15:41:17,560 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-4) WFLYUT0007: Undertow HTTP listener default stopped, was bound to [::1]:8080
 [0m [0m15:41:17,560 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-4) WFLYSRV0028: Stopped deployment arquillian-service (runtime-name: arquillian-service) in 12ms
 [0m [0m15:41:17,560 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-4) WFLYUT0004: Undertow 1.3.16.Final stopping
 [0m [0m15:41:17,571 INFO  [org.jboss.as.connector.deployers.jdbc] (MSC service thread 1-2) WFLYJCA0019: Stopped Driver service with driver-name = h2
 [0m [0m15:41:17,577 INFO  [org.jboss.as.mail.extension] (MSC service thread 1-2) WFLYMAIL0002: Unbound mail session [java:jboss/mail/Default]
 [0m [0m15:41:17,579 INFO  [org.jboss.as] (MSC service thread 1-1) WFLYSRV0050: WildFly Full 10.1.0.Final-SNAPSHOT (WildFly Core 2.1.0.CR2) stopped in 30ms
 [0m [0m15:41:17,579 INFO  [org.jboss.as] (MSC service thread 1-1) WFLYSRV0049: WildFly Full 10.1.0.Final-SNAPSHOT (WildFly Core 2.1.0.CR2) starting
 [0m [0m15:41:17,630 INFO  [org.jboss.as.server] (Controller Boot Thread) WFLYSRV0039: Creating http management service using socket-binding (management-http)
 [0m [0m15:41:17,632 INFO  [org.jboss.as.connector.subsystems.datasources] (ServerService Thread Pool -- 32) WFLYJCA0004: Deploying JDBC-compliant driver class org.h2.Driver (version 1.3)
 [0m [0m15:41:17,641 INFO  [org.jboss.as.clustering.infinispan] (ServerService Thread Pool -- 38) WFLYCLINF0001: Activating Infinispan subsystem.
 [0m [0m15:41:17,641 INFO  [org.wildfly.extension.io] (ServerService Thread Pool -- 37) WFLYIO001: Worker 'default' has auto-configured to 4 core threads with 32 task threads based on your 2 available processors
 [0m [0m15:41:17,642 INFO  [org.jboss.as.ejb3] (MSC service thread 1-2) WFLYEJB0481: Strict pool slsb-strict-max-pool is using a max instance size of 32 (per class), which is derived from thread worker pool sizing.
 [0m [0m15:41:17,642 INFO  [org.jboss.as.ejb3] (MSC service thread 1-2) WFLYEJB0482: Strict pool mdb-strict-max-pool is using a max instance size of 8 (per class), which is derived from the number of CPUs on this host.
 [0m [0m15:41:17,642 INFO  [org.jboss.as.connector] (MSC service thread 1-2) WFLYJCA0009: Starting JCA Subsystem (WildFly/IronJacamar 1.3.3.Final)
 [0m [0m15:41:17,642 INFO  [org.jboss.as.connector.deployers.jdbc] (MSC service thread 1-3) WFLYJCA0018: Started Driver service with driver-name = h2
 [0m [0m15:41:17,659 INFO  [org.jboss.as.naming] (ServerService Thread Pool -- 46) WFLYNAM0001: Activating Naming Subsystem
 [0m [0m15:41:17,660 INFO  [org.jboss.as.naming] (MSC service thread 1-3) WFLYNAM0003: Starting Naming Service
 [0m [0m15:41:17,660 INFO  [org.jboss.as.mail.extension] (MSC service thread 1-3) WFLYMAIL0001: Bound mail session [java:jboss/mail/Default]
 [0m [0m15:41:17,661 INFO  [org.jboss.as.security] (ServerService Thread Pool -- 53) WFLYSEC0002: Activating Security Subsystem
 [0m [33m15:41:17,663 WARN  [org.jboss.as.txn] (ServerService Thread Pool -- 54) WFLYTX0013: Node identifier property is set to the default value. Please make sure it is unique.
 [0m [0m15:41:17,663 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 55) WFLYUT0003: Undertow 1.3.16.Final starting
 [0m [0m15:41:17,664 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 55) WFLYUT0014: Creating file handler for path '/opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/basic/target/jbossas/welcome-content' with options [directory-listing: 'false', follow-symlink: 'false', case-sensitive: 'true', safe-symlink-paths: '[]']
 [0m [0m15:41:17,670 INFO  [org.jboss.as.webservices] (ServerService Thread Pool -- 56) WFLYWS0002: Activating WebServices Extension
 [0m [0m15:41:17,672 INFO  [org.jboss.as.security] (MSC service thread 1-4) WFLYSEC0001: Current PicketBox version=4.9.4.Final
 [0m [0m15:41:17,673 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-4) WFLYUT0003: Undertow 1.3.16.Final starting
 [0m [0m15:41:17,674 INFO  [org.jboss.as.server.deployment.scanner] (MSC service thread 1-4) WFLYDS0013: Started FileSystemDeploymentService for directory /opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/basic/target/jbossas/standalone/deployments
 [0m [0m15:41:17,674 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-4) WFLYSRV0027: Starting deployment of "arquillian-service" (runtime-name: "arquillian-service")
 [0m [0m15:41:17,743 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-4) WFLYUT0012: Started server default-server.
 [0m [0m15:41:17,743 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-1) WFLYUT0018: Host default-host starting
 [0m [0m15:41:17,744 INFO  [org.jboss.as.connector.subsystems.datasources] (MSC service thread 1-2) WFLYJCA0001: Bound data source [java:jboss/datasources/ExampleDS]
 [0m [0m15:41:17,744 INFO  [org.wildfly.extension.undertow] (MSC service thread 1-4) WFLYUT0006: Undertow HTTP listener default listening on [::1]:8080
 [0m [0m15:41:17,745 INFO  [org.jboss.ws.common.management] (MSC service thread 1-4) JBWS022052: Starting JBossWS 5.1.3.SP1 (Apache CXF 3.1.4) 
 [0m [33m15:41:17,748 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-4) WFLYSRV0018: Deployment "deployment.arquillian-service" is using a private module ("org.jboss.as.jmx:main") which may be changed or removed in future versions without notice.
 [0m [33m15:41:17,749 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-4) WFLYSRV0018: Deployment "deployment.arquillian-service" is using a private module ("org.jboss.as.server:main") which may be changed or removed in future versions without notice.
 [0m [33m15:41:17,749 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-4) WFLYSRV0018: Deployment "deployment.arquillian-service" is using a private module ("org.jboss.jandex:main") which may be changed or removed in future versions without notice.
 [0m [33m15:41:17,749 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-4) WFLYSRV0018: Deployment "deployment.arquillian-service" is using a private module ("org.wildfly.security.manager:main") which may be changed or removed in future versions without notice.
 [0m [33m15:41:17,754 WARN  [org.jboss.weld.deployer] (MSC service thread 1-4) WFLYWELD0013: Deployment deployment "arquillian-service" contains CDI annotations but no bean archive was not found. (No beans.xml nor class with bean defining annotations)


```

--- a/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/response/DefaultResponseCodeAtRootTestCase.java
+++ b/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/response/DefaultResponseCodeAtRootTestCase.java
@@ -107,9 +107,9 @@ public class DefaultResponseCodeAtRootTestCase extends ContainerResourceMgmtTest
         operation.get("value").set(506);
         cob.addStep(operation);
         // if location service is removed, if no deployment == no virtual host.
-//        operation = createOpNode("subsystem=undertow/server=default-server/host=default-host", "remove");
-//        operation.get("address").add("location","/");
-//        cob.addStep(operation);
+        operation = createOpNode("subsystem=undertow/server=default-server/host=default-host", "remove");
+        operation.get("address").add("location","/");
+        cob.addStep(operation);
         executeOperation(cob.build().getOperation());
         executeReloadAndWaitForCompletion(getModelControllerClient());
         deployer.deploy("test");
@@ -128,10 +128,10 @@ public class DefaultResponseCodeAtRootTestCase extends ContainerResourceMgmtTest
             operation = createOpNode("subsystem=undertow/server=default-server/host=default-host", "undefine-attribute");
             operation.get("name").set("default-response-code");
             cob.addStep(operation);
-//            operation = createOpNode("subsystem=undertow/server=default-server/host=default-host", "add");
-//            operation.get("address").add("location","/");
-//            operation.get("handler").set("welcome-content");
-//            cob.addStep(operation);
+            operation = createOpNode("subsystem=undertow/server=default-server/host=default-host", "add");
+            operation.get("address").add("location","/");
+            operation.get("handler").set("welcome-content");
+            cob.addStep(operation);
             executeOperation(cob.build().getOperation());
             executeReloadAndWaitForCompletion(getModelControllerClient());
         }

--- a/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/response/RewriteLocationByDeplomentTestCase.java
+++ b/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/response/RewriteLocationByDeplomentTestCase.java
@@ -1,0 +1,101 @@
+package org.jboss.as.test.integration.web.response;
+
+import java.io.IOException;
+import java.net.URI;
+
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.jboss.arquillian.container.test.api.Deployer;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.container.test.api.RunAsClient;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.as.arquillian.api.ContainerResource;
+import org.jboss.as.arquillian.container.ManagementClient;
+import org.jboss.as.test.integration.management.base.ContainerResourceMgmtTestBase;
+import org.jboss.as.test.integration.management.util.MgmtOperationException;
+import org.jboss.as.test.integration.management.util.ModelUtil;
+import org.jboss.as.test.shared.ServerReload;
+import org.jboss.dmr.ModelNode;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Tomas Hofman (thofman@redhat.com)
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+public class RewriteLocationByDeplomentTestCase extends ContainerResourceMgmtTestBase {
+
+    @ArquillianResource
+    Deployer deployer;
+
+    @SuppressWarnings("WeakerAccess")
+    @ContainerResource
+    ManagementClient managementClient;
+
+    @Deployment(name = "app", managed = false)
+    public static WebArchive deployment() {
+        WebArchive war = ShrinkWrap.create(WebArchive.class, DefaultResponseCodeTestCase.class.getSimpleName() + ".war");
+        war.addClasses(SimpleRootServlet.class);
+        war.addAsWebInfResource(new StringAsset("<jboss-web><context-root>/test</context-root></jboss-web>"),
+                "jboss-web.xml");
+        return war;
+    }
+
+    @Test
+    public void testDeploymentOverLocation() throws IOException, MgmtOperationException {
+        // check that "/test" path returns 404
+        HttpResponse response = getResponse("/test");
+        Assert.assertEquals(404, response.getStatusLine().getStatusCode());
+
+        // create location "/test" serving welcome-content
+        ModelNode op = ModelUtil.createOpNode("subsystem=undertow/server=default-server/host=default-host", "add");
+        op.get("address").add("location", "/test");
+        op.get("handler").set("welcome-content");
+        executeOperation(op);
+
+        try {
+            ServerReload.executeReloadAndWaitForCompletion(getModelControllerClient());
+            response = getResponse("/test");
+            Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+            Assert.assertTrue("Expected to receive welcome page, but content length is 0",
+                    response.getEntity().getContentLength() > 0);
+
+            // deploy an app at the same location and check it's accessible
+            deployer.deploy("app");
+            response = getResponse("/test");
+            Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+            Assert.assertTrue("Expected to receive servlet response without any content",
+                    response.getEntity().getContentLength() == 0);
+
+            // undeploy app and check that welcome-content is accessible again
+            deployer.undeploy("app");
+            response = getResponse("/test");
+            Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+            Assert.assertTrue("Expected to receive welcome page, but content length is 0",
+                    response.getEntity().getContentLength() > 0);
+        } finally {
+            op = ModelUtil.createOpNode("subsystem=undertow/server=default-server/host=default-host", "remove");
+            op.get("address").add("location", "/test");
+            //noinspection ThrowFromFinallyBlock
+            executeOperation(op);
+            ServerReload.executeReloadAndWaitForCompletion(getModelControllerClient());
+
+            //noinspection ThrowFromFinallyBlock
+            response = getResponse("/test");
+            Assert.assertEquals(404, response.getStatusLine().getStatusCode());
+        }
+    }
+
+    private HttpResponse getResponse(String path) throws IOException {
+        URI webUri = managementClient.getWebUri();
+        HttpGet httpGet = new HttpGet(webUri.resolve(path));
+        return HttpClientBuilder.create().build().execute(httpGet);
+    }
+}

--- a/undertow/src/main/java/org/wildfly/extension/undertow/Host.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/Host.java
@@ -40,6 +40,7 @@ import io.undertow.server.HttpServerExchange;
 import io.undertow.server.handlers.PathHandler;
 import io.undertow.servlet.api.Deployment;
 import io.undertow.servlet.api.DeploymentInfo;
+import io.undertow.util.CopyOnWriteMap;
 import io.undertow.util.Methods;
 import org.jboss.msc.service.Service;
 import org.jboss.msc.service.StartContext;
@@ -64,6 +65,7 @@ public class Host implements Service<Host>, FilterLocation {
     private volatile AccessLogService  accessLogService;
     private final List<FilterRef> filters = new CopyOnWriteArrayList<>();
     private final Set<Deployment> deployments = new CopyOnWriteArraySet<>();
+    private final Map<String, LocationService> locations = new CopyOnWriteMap<>();
     private final Map<String, AuthenticationMechanism> additionalAuthenticationMechanisms = new ConcurrentHashMap<>();
     private final HostRootHandler hostRootHandler = new HostRootHandler();
 
@@ -213,9 +215,25 @@ public class Host implements Service<Host>, FilterLocation {
 
     public void unregisterHandler(String path) {
         pathHandler.removePrefixPath(path);
-        if(path.equals("/")){
+        // if there is registered location for given path, serve it from now on
+        LocationService location = locations.get(path);
+        if (location != null) {
+            pathHandler.addPrefixPath(location.getLocationPath(), location.getLocationHandler());
+        }
+        // else serve the default response code
+        else if (path.equals("/")) {
             this.setupDefaultResponseCodeHandler();
         }
+    }
+
+    void registerLocation(LocationService location) {
+        locations.put(location.getLocationPath(), location);
+        registerHandler(location.getLocationPath(), location.getLocationHandler());
+    }
+
+    void unregisterLocation(LocationService location) {
+        locations.remove(location.getLocationPath());
+        unregisterHandler(location.getLocationPath());
     }
 
     /**

--- a/undertow/src/main/java/org/wildfly/extension/undertow/LocationService.java
+++ b/undertow/src/main/java/org/wildfly/extension/undertow/LocationService.java
@@ -57,12 +57,12 @@ public class LocationService implements Service<LocationService>, FilterLocation
     @Override
     public void start(StartContext context) throws StartException {
         UndertowLogger.ROOT_LOGGER.tracef("registering handler %s under path '%s'", httpHandler.getValue(), locationPath);
-        host.getValue().registerHandler(locationPath, locationHandler);
+        host.getValue().registerLocation(this);
     }
 
     @Override
     public void stop(StopContext context) {
-        host.getValue().unregisterHandler(locationPath);
+        host.getValue().unregisterLocation(this);
     }
 
     @Override
@@ -76,6 +76,14 @@ public class LocationService implements Service<LocationService>, FilterLocation
 
     InjectedValue<HttpHandler> getHttpHandler() {
         return httpHandler;
+    }
+
+    String getLocationPath() {
+        return locationPath;
+    }
+
+    LocationHandler getLocationHandler() {
+        return locationHandler;
     }
 
     private HttpHandler configureHandler() {

